### PR TITLE
add Hasura Graphql Engine API type-safe interfaces

### PR DESCRIPTION
**Pull request by hgiasac**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test YOUR_PACKAGE_NAME`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/hasura/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+export type JSONColumnBoolExp = ColumnBoolExp<string>;
+export interface JSONBColumnBoolExp extends ColumnBoolExp<string> {
+    _contains?: string;
+    _contained_in?: string;
+    _has_key?: string;
+    _has_keys_all?: string;
+    _has_keys_any?: string;
+}
+
+type Equals<X, Y> = (<T>() => T extends X ? 1 : 2) extends <T>() => T extends Y ? 1 : 2 ? true : false;
```
Comment (amcasey) [inline types/hasura/index.d.ts:41]: I think this is usually defined as
```ts
type Equals<X, Y> = [X] extends [Y] ? [Y] extends [X] ? true : false : false;
```

Then you don't need to suppress the lint rule.

File: types/hasura/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+type ScalarType = string | number | boolean | ScalarJSON<unknown> | ScalarJSONB<unknown>;
+type ObjectType = Record<string, ScalarType | Record<string, ScalarType> | Array<Record<string, ScalarType>>>;
+export type ScalarJSON<T> = T & {
+    __type?: 'json';
```
Comment (amcasey) [inline types/hasura/index.d.ts:45]: This looks like it was intended to be a brand and brands aren't usually optional.  Am I missing something?

**Review by amcasey — CHANGES_REQUESTED** [review]

File: types/hasura/index.d.ts (hunk 2 above)
Comment (hgiasac) [reply types/hasura/index.d.ts:45]: @amcasey My idea is defining a wrapped type from original type `<T>`, so the Typescript's condition check can know it is JSON type. 

I tried to define 
```
type ScalarJSON<T> = T;
```
However the Equals check doesn't work. It still think 2 types are the same
```ts
Equals<T, ScalarJSON<T>> // true
```
So I decide adding an optional field to make them different. The user doesn't need to know or input this field.

However, if you have better idea, I'm really appreciated to know

File: types/hasura/index.d.ts (hunk 1 above)
Comment (hgiasac) [reply types/hasura/index.d.ts:41]: I tried your suggestion, however, below equality check don't work anymore
```ts
export type ScalarJSON<T> = T & {
    __type?: 'json';
}

hasura.Equals<{ a: string }, hasura.ScalarJSON<{ a: string }>> // true
```

**Review by amcasey — APPROVED** [review]
Seems like you've thought things through.  If you're happy, I'm happy.

**Comment by hgiasac** [comment]
Ready to merge

**Pull request merged**